syntax = "proto3";
import "google/protobuf/timestamp.proto";
import "FileInfo.proto";
import "TrackLayoutMetadata.proto";

package IRacingTools.Models.Session;

// Enum representing the sub-types of a session. This distinguishes
// specific parts of a session, such as practice, qualifying, or race phases.
enum SessionSubType {
  SESSION_SUB_TYPE_UNKNOWN = 0;
  SESSION_SUB_TYPE_PRACTICE = 1;
  SESSION_SUB_TYPE_QUALIFY = 2;
  SESSION_SUB_TYPE_RACE = 3;
}

enum SessionSubTimingType {
  SESSION_SUB_TIMING_TYPE_UNKNOWN = 0;
  SESSION_SUB_TIMING_TYPE_TIMED = 1;
  SESSION_SUB_TIMING_TYPE_LAPS = 2;
}

message SessionTiming {
  bool is_live = 1;
  bool is_valid = 2;

  int32 sample_index = 50;
  int32 sample_count = 51;

  int32 ticks = 60;
  int32 tick_count = 61;

  int32 session_sub_count = 70;
  int32 session_sub_num = 71;
  SessionSubType session_sub_type = 72;
  SessionSubTimingType session_sub_timing_type = 73;

  int32 session_sub_lap = 75;
  int32 session_sub_lap_remaining = 76;
  int32 session_sub_lap_count = 77;
  int32 session_sub_time = 80;
  int32 session_sub_time_remaining = 81;
  int32 session_sub_time_total = 82;

}

enum SessionType {
  SESSION_TYPE_LIVE = 0;
  SESSION_TYPE_DISK = 1;
}

enum SessionCarStatus {
  SESSION_CAR_STATUS_INVALID = 0;
  SESSION_CAR_STATUS_GET_IN_CAR = 1;
  SESSION_CAR_STATUS_WARMUP = 3;
  SESSION_CAR_STATUS_PARADE_LAPS = 4;
  SESSION_CAR_STATUS_RACE = 5;
  SESSION_CAR_STATUS_CHECKERED = 6;
  SESSION_CAR_STATUS_COOLDOWN = 7;
}


enum SessionStatus {
  SESSION_STATUS_CREATED = 0;
  SESSION_STATUS_READY = 1;
  SESSION_STATUS_PAUSED = 2;
  SESSION_STATUS_RUNNING = 5;
  SESSION_STATUS_ERROR = 10;
}

message SessionData {

  string id = 1;
  SessionType type = 2;

  SessionStatus status = 5;

  int32 sub_count = 6;
  int32 sub_id = 7;
  int32 sub_num = 8;
  SessionSubType sub_type = 9;

  SessionTiming timing = 15;

  TrackLayoutMetadata track_layout_metadata = 20;
  FileInfo file_info = 21;

  SessionCarStatus car_status = 30;

  string session_info_json = 90;
  string session_info_yaml = 91;
}
